File: Binary Search/BinaryTree.java
package BinaryTree;
import java.util.*;
class Node{
    int key ; 
    Node left , right ; 
    Node(int key){
        this.key = key;
        this.left = null ;
        this.right = null;
    }
}
public class BinaryTree {
    static void preorderDFS(Node node){
        if (node == null){
            return;
        }
        System.out.print(node.key + " ");
        preorderDFS(node.left);
        preorderDFS(node.right);
    }
    static void inorderDFS(Node node){
        if (node == null){
            return;
        }
        inorderDFS(node.left);
        System.out.print(node.key + " ");
        inorderDFS(node.right);
    }
    static void postOrderDFS(Node node){
        if (node == null){
            return;
        }
        postOrderDFS(node.left);
        postOrderDFS(node.right);
        System.out.print(node.key + " ");
    }
    static void bfs(Node root){
        if (root == null){
            return;
        }
        Queue<Node> queue = new LinkedList<>();
        queue.add(root);
        while (!queue.isEmpty()) {
            Node node = queue.poll();
            System.out.print(node.key + " ");
            if (node.left != null) {
                queue.add(node.left);
            }
            if (node.right != null){
                queue.add(node.right);
            }
        }
    }
    static Node add_binary_tree(Node root , int key){
        if (root == null)
            return new Node(key);
        Queue<Node> queue = new LinkedList<>();
        queue.add(root);
        while (!queue.isEmpty()){
            Node temp = queue.poll();
            if (temp.left == null){
                temp.left = new Node(key);
                break;
            }else{
                queue.add(temp.left);
            }
            if (temp.right == null){
                temp.right = new Node(key);
                break;
            }else{
                queue.add(temp.right);
            }
        }
        return root;
    }
    static boolean searchDFS(Node root , int x){
        if (root == null)
            return false;
        if (root.key == x )
            return true; 
        boolean left_res = searchDFS(root.left, x);
        boolean right_res = searchDFS(root.right, x);
        return left_res || right_res;
    }
    static Node delete(Node root , int x ){
        if (root == null)
            return null;
        Queue<Node> queue = new LinkedList<>();
        Node target = null;
        queue.add(root);
        while(!queue.isEmpty()){
            Node current = queue.poll();
            if (current.key == x){
                target = current;
                break;
            }
            if (current.left != null)
                queue.add(current.left);
            if (current.right != null)
                queue.add(current.right);
        }
        if (target == null)
            return root;
        Node lastNode  = null;
        Node lastParent = null;
        Queue<Node> queue2 = new LinkedList<>();
        Queue<Node> parentQueue = new LinkedList<>();
        queue2.add(root);
        parentQueue.add(null);
        while(!queue2.isEmpty()){
            Node current = queue2.poll();
            Node parent = parentQueue.poll();

            lastNode = current;
            lastParent = parent ;

            if (current.left != null)
                queue2.add(current.left);
                parentQueue.add(current);
            if (current.right != null)
                queue2.add(current.right);
                parentQueue.add(current);
        }
        target.key = lastNode.key;
        if (lastParent != null){
            if (lastParent.left != lastNode){
                lastParent.left = null;
            }else{
                lastParent.right = null;
            }
        }else{
            return null;
        }
        return root;
    } 
     public static void main(String[] args) {
        Node root1 = new Node(78);
        root1.left = new Node(62);
        root1.right = new Node(53);
        root1.left.left = new Node(34);
        root1.left.right = new Node(25);
        System.out.println("Duyet truoc ");
        preorderDFS(root1);

        System.out.println("\nDuyet theo thu tu tu truoc den sau ");
        inorderDFS(root1);

        System.out.println("\nDuyet theo thu tu tu sau vao truoc ");
        postOrderDFS(root1);

        int key = 6 ; 
        root1 = add_binary_tree(root1, key);
        System.out.println("\nSau khi them ");
        inorderDFS(root1);

        key = 10 ; 
        root1 = add_binary_tree(root1, key);
        System.out.println("\nSau khi them ");
        inorderDFS(root1);
        
        key = 14 ; 
        root1 = add_binary_tree(root1, key);
        System.out.println("\nSau khi them ");
        inorderDFS(root1);
        System.out.println();

        int val  = 25;
        delete(root1, val);
        System.out.println("Xoa phan tu trong cay nhi phan");
        inorderDFS(root1);
        System.out.println();

        int x = 6; 
        if (searchDFS(root1, x))
            System.out.println(x + " Ton trong cay nhi phan");
        else 
            System.out.println(x + " Khong ton tai trong cay nhi phan");
        
        
    }
}

File: Binary Search/BinaryTree.py
from collections import deque
class Node:
    def __init__(self, data):
        self.data = data  
        self.left = self.right = None 

    def preorderDFS(self,node):
        if node is None :
            return 
        print(node.data , end = " ")
        self.preorderDFS(node.left)
        self.preorderDFS(node.right)
    
    def inorderDFS(self,node):
        if node is None : 
            return 
        self.inorderDFS(node.left)
        print(node.data , end = " ")
        self.inorderDFS(node.right)
    
    def postorderDFS(self,node):
        if node is None :
            return
        self.postorderDFS(node.left)
        self.postorderDFS(node.right)
        print(node.data , end = " ")

    def bfs(root):
        if root is None:
            return
        queue = [root]
        while queue:
            node = queue.pop(0)
            print(node.data, end=' ')
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)
    def add_binary_tree(root,data):
        if root is None : 
            return Node(data)
        queue = deque([root])
        queue.append(root)
        while len(queue) > 0 : 
            temp = queue.popleft()
            if temp.left is None : 
                temp.left = Node(data)
                break
            else:
                queue.append(temp.left)
            if temp.right is None : 
                temp.right = Node(data)
                break
            else:
                queue.append(temp.right)
        return root
    def searchDFS(self , root , x ):
            if root is None :
                return False
            if root.data == x :
                return True 
            left_res = self.searchDFS(root.left , x);
            right_res = self.searchDFS(root.right , x )
            return left_res or right_res
    def delete(self,root , x ):
        if root is None :
            return None 
        queue = deque([root])
        target = None 
        while queue:
            current = queue.popleft()
            if current.data == x:
                target = current 
                break
            if current.left :
                queue.append(current.left)
            if current.right :
                queue.append(current.right)
        if target is None :
            return root 
        lastNode = None 
        lastParent = None 
        queue = deque([(root, None)])
        while queue : 
            current,parent=queue.popleft()
            lastNode = current 
            lastParent = parent
            if current.left :
                queue.append((current.left , current))
            if current.right :
                queue.append((current.right , current))
        target.data = lastNode.data
        if lastParent :
            if lastParent.left == lastNode:
                lastParent.left = None 
            else:
                lastParent.right = None 
        else:
            return None 
        return root
if __name__ == '__main__':
    root = Node(84)
    root.left = Node(32)
    root.right = Node(21)
    root.left.left  = Node(48)
    root.left.right = Node(74)

    print("Duyet truoc")
    root.preorderDFS(root)

    print("\nDuyet theo thu tu tu truoc ra sau " )
    root.inorderDFS(root)
    
    print("\nDuyet theo thu tu tu sau ra truoc " )
    root.postorderDFS(root)

    key = 83
    root = Node.add_binary_tree(root, key)
    print("\nSau khi them :")
    root.inorderDFS(root)

    key = 95
    root = Node.add_binary_tree(root,key)
    print("\nSau khi chen : ")
    root.inorderDFS(root)

    x = 48
    if (root.searchDFS(root,x)):
        print(f"\n{x} ton tai trong cay nhi phan")
    else:
        print(f"\n{x} khong ton tai trong cay nhi phan")
    
    x = 48
    root = root.delete(root , x )
    print(f"Sau khi xoa {x}")
    root.inorderDFS(root)

    x = 48
    if (root.searchDFS(root,x)):
        print(f"\n{x} ton tai trong cay nhi phan")
    else:
        print(f"\n{x} khong ton tai trong cay nhi phan")
    


    
